nixos/modules/services/hardware/usbipd.nix

```diff
@@ -0,0 +1,99 @@
+{
+  lib,
+  pkgs,
+  config,
+  ...
+}:
+let
+  cfg = config.services.usbipd;
+  device = lib.types.submodule {
+    options = {
+      productid = lib.mkOption {
+        type = lib.types.str;
+        description = "The product id of the device";
+      };
+      vendorid = lib.mkOption {
+        type = lib.types.str;
+        description = "The vendor id of the device";
+      };
+    };
+  };
+in
+{
+  options.services.usbipd = {
+    enable = lib.mkEnableOption "usbip server";
+    kernelPackage = lib.mkPackageOption pkgs.linuxPackages_latest "usbip" { };
+    devices = lib.mkOption {
+      type = lib.types.listOf device;
+      default = [ ];
+      description = "List of USB devices to watch and automatically export.";
+      example = [
+        {
+          productid = "xxxx";
+          vendorid = "xxxx";
+        }
+      ];
+    };
+    openFirewall = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = "Open port 3240 for usbipd";
+      example = false;
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    boot.extraModulePackages = [ cfg.kernelPackage ];
+    boot.kernelModules = [
+      "usbip-core"
+      "usbip-host"
+    ];
+    networking.firewall.allowedTCPPorts = lib.mkIf cfg.openFirewall [ 3240 ];
+    services.udev.extraRules = lib.strings.concatLines (
+      (map (
+        dev:
+        "ACTION==\"add\", SUBSYSTEM==\"usb\", ATTRS{idProduct}==\"${dev.productid}\", ATTRS{idVendor}==\"${dev.vendorid}\", RUN+=\"${pkgs.systemd}/bin/systemctl restart usbip-${dev.vendorid}:${dev.productid}.service\""
+      ) cfg.devices)
+    );
+
+    systemd.services =
+      (builtins.listToAttrs (
+        map (dev: {
+          name = "usbip-${dev.vendorid}:${dev.productid}";
+          value = {
+            after = [ "usbipd.service" ];
+            script = ''
+              set +e
+              devices=$(${cfg.kernelPackage}/bin/usbip list -l | grep -E "^.*- busid.*(${dev.vendorid}:${dev.productid})" )
+              echo $devices | while read device; do
+                output=$(${cfg.kernelPackage}/bin/usbip -d bind -b $(echo $device | ${pkgs.gawk}/bin/awk '{ print $3 }') 2>&1)
+                code=$?
+
+                echo $output
+                if [[ $output =~ "already bound" ]] || [ $code -eq 0 ]; then
+                  continue
+                else
+                  exit $code
+                fi
+              done
+            '';
+            serviceConfig.Type = "oneshot";
+            serviceConfig.RemainAfterExit = true;
+            restartTriggers = [ "usbipd.service" ];
+          };
+        }) cfg.devices
+      ))
+      // {
+        usbipd = {
+          wantedBy = [ "multi-user.target" ];
+          script = ''
+            ${lib.getExe' pkgs.kmod "modprobe"} usbip-host usbip-core
```

Comment:
This is done via boot.kernelModules already, I think?

Reply:
It is not, e.g. rke2 does this and several others.

Reply:
rke2 does not set boot.kernelModules. My understanding is that putting something in `boot.kernelModules` is almost equivalent to a `modprobe` call at boot time.

Reply:
If I could get confirmation, I am happy to remove it. Though when I initially tested this case the module was not loaded until after reboot, so that is why I added it.

Reply:
Looking at kernel.nix it seems like it is supposed to autoload, but in my experience it didn’t and the wording seems to indicate otherwise “The set of kernel modules to be loaded in the second stage of the boot process”. In any case I will remove it.

Reply:
> Though when I initially tested this case the module was not loaded until after reboot, so that is why I added it.

That is correct. It will not work with a simple `nixos-rebuild switch`, only a reboot will apply a change to `boot.kernelModules`.